clients/drcachesim/tracer/output.cpp

```diff
@@ -989,9 +1002,26 @@ process_and_output_buffer(void *drcontext, bool skip_size_cap)
         if (op_offline.get_value() && op_split_windows.get_value())
             buf_ptr += instru->append_thread_exit(buf_ptr, dr_get_thread_id(drcontext));
     }
-
-    uintptr_t mode = tracing_mode.load(std::memory_order_acquire);
-
+    // Switch to instruction-tracing mode by adding FILTER_ENDPOINT marker if another
+    // thread triggered the switch.
+    ptr_int_t mode = tracing_mode.load(std::memory_order_acquire);
+    if (get_local_mode(data) != mode) {
+        if (get_local_mode(data) == BBDUP_MODE_L0_FILTER) {
+            NOTIFY(0, "Thread %d: filter mode changed\n", dr_get_thread_id(drcontext));
+            // If a switch occurred, then it is possible that the buffer
+            // contains a mix of filtered and unfiltered records. We discard
+            // the buffer by resetting the buffer pointer.
```

Comment:
Actually, can we identify the transition point in raw2trace by the presence of a PC record of size 0 (if I'm remembering that right) for filtered?  So we don't need to discard anything, and we can have raw2trace insert this FILTER_ENDPOINT marker for us at the right spot?

Update: OK I made it to tracer.cpp and I see you've added a check for the mode changing at the bottom of every block.  So at most the most recent block might contain full trace data: it is not possible for a prior block to contain full data, right?  So don't we just need to identify whether this block has filtered or unfiltered, and if unfiltered we insert the marker before its own data, and if filtered we insert it after, and we never need to discard any data?

Reply:
> Actually, can we identify the transition point in raw2trace by the presence of a PC record of size 0 (if I'm remembering that right) for filtered? So we don't need to discard anything, and we can have raw2trace insert this FILTER_ENDPOINT marker for us at the right spot?

In filtered traces, we can have set of entries of the form (1) or (2) whereas for unfiltered traces, we have entries of the form (3) as shown below. I am not sure how to distinguish between (1) and (3). Therefore it will be difficult for raw2trace to add the marker.

1. PC: 1 instruction
2. PC: 0 instructions, 1 MEMREF
3. PC: >0 instructions, >=0 MEMREF

>  So don't we just need to identify whether this block has filtered or unfiltered, and if unfiltered we insert the marker before its own data, and if filtered we insert it after, and we never need to discard any data?

It is also possible that the buffer has a mix of filtered and unfiltered records, right? Then we have the same problem as above.

Reply:
> I am not sure how to distinguish between (1) and (3)

Right, agreed.

However, only the most recent basic block can have unfiltered data.  Once the mode switch is made, it will take effect in some thread at the top of a block in the drbbdup mode dispatch.  Then at the bottom of that block it will hit the new check and enter the clean call.  So if we walk backward to the first PC entry we find (since unfiltered has just one PC at the start) that must be the transition point, right?

Reply:
I have made this change.